### ISSUE TITLE
Must be superuser to alter superusers

### ISSUE THREAD

**Issue by techpet** [kartoza/docker-postgis#373]
While restarting my postgis container after a power failure, the database server never got up again. These are the logged messages:

```
postgis_prod    | 2022-05-03 17:23:36.216 UTC [36] LOG:  starting PostgreSQL 12.4 (Debian 1
2.4-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
postgis_prod    | 2022-05-03 17:23:36.216 UTC [36] LOG:  listening on IPv4 address "127.0.0
.1", port 5432
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | 2022-05-03 17:23:36.233 UTC [36] LOG:  listening on Unix socket "/var/run
/postgresql/.s.PGSQL.5432"
postgis_prod    | 2022-05-03 17:23:36.275 UTC [41] LOG:  database system was interrupted; l
ast known up at 2022-05-03 17:23:32 UTC
postgis_prod    | 2022-05-03 17:23:36.301 UTC [41] LOG:  database system was not properly s
hut down; automatic recovery in progress
postgis_prod    | 2022-05-03 17:23:36.314 UTC [41] LOG:  invalid record length at 2/7011385
0: wanted 24, got 0
postgis_prod    | 2022-05-03 17:23:36.314 UTC [41] LOG:  redo is not required
postgis_prod    | 2022-05-03 17:23:36.390 UTC [36] LOG:  database system is ready to accept
 connections
postgis_prod    | 2022-05-03 17:23:36.395 UTC [47] LOG:  pg_cron scheduler started
postgis_prod    | /var/run/postgresql:5432 - accepting connections
postgis_prod    | postgres ready
postgis_prod    | Setup postgres User:Password
postgis_prod    | Creating superuser myuser
postgis_prod    | 2022-05-03 17:23:37.472 UTC [84] postgres@postgres ERROR:  must be superu
ser to alter superusers
postgis_prod    | 2022-05-03 17:23:37.472 UTC [84] postgres@postgres STATEMENT:  ALTER USER
 myuser WITH SUPERUSER ENCRYPTED PASSWORD 'mypass';
postgis_prod    | ERROR:  must be superuser to alter superusers
postgis_prod    | Setup master database
postgis_prod    | 2022-05-03 17:23:41.669 UTC [34] LOG:  starting PostgreSQL 12.4 (Debian 1
2.4-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
postgis_prod    | 2022-05-03 17:23:41.669 UTC [34] LOG:  listening on IPv4 address "127.0.0
.1", port 5432
postgis_prod    | 2022-05-03 17:23:41.675 UTC [34] FATAL:  lock file "/var/run/postgresql/.
s.PGSQL.5432.lock" already exists
postgis_prod    | 2022-05-03 17:23:41.675 UTC [34] HINT:  Is another postmaster (PID 36) us
ing socket file "/var/run/postgresql/.s.PGSQL.5432"?
postgis_prod    | 2022-05-03 17:23:41.675 UTC [34] LOG:  database system is shut down
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
postgis_prod    | /var/run/postgresql:5432 - no response
```
Seems like my defined superuser somehow lost his superuser permissions, but how is this possible? How could I alter the permissions again if I cannot even connect to the server?

**Comment by NyakudyaA** [comment]
https://stackoverflow.com/questions/66159502/postgresql-unable-to-alter-role-to-superuser

It looks like the postgres user is no longer available in your setup. I suggest you add it manually because all the initialization scripts runs using the postgres user

**Comment by techpet** [comment]
I've accessed a bash shell in the container and I've tried the solution proposed but I keep getting:
```
root@e5099e27bdbf:/# su - postgres
postgres@e5099e27bdbf:~$ psql -U postgres
psql: error: could not connect to server: could not connect to server: Connection refused
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
```

postgresql service inside the container is down and even after restarting it, it always stops so I don't have a way of connecting to it. 

If this is of any use, I am using volumes to persist data on host.

**Comment by techpet** [comment]
Finally managed to solve this and I am writing down the process in case anyone needs it in the future.

Indeed, the problem was the one indicated by @NyakudyaA but the solution needed a twist to the answer provided in the stackoverflow link. 

At first I had to access the container as postgres user with a bash shell. Afterwards, as the postgres server kept shuting down, pgsql was of no use and I had to start the PostgreSQL “stand-alone backend” with which a single user can interact directly without using a separate client. In order to achieve this, I had to locate the postgres binary path (`/usr/lib/postgresql/12/bin/postgres` in my case) and the data path (`/var/lib/postgresql/12/main/` in my case) and also specify the database to connect to (`mydatabase`).
```
postgres@85a2f96c90ed:/$ /usr/lib/postgresql/12/bin/postgres --single -D /var/lib/postgresql/12/main/ mydatabase
```

Finally, I specified that postgres user is superuser by executing:
```
ALTER USER postgres SUPERUSER;
```
and when I started postgresql server again everything run smoothly this time.
